Add `shouldTriggerDismiss` callback parameter to the `Dismissible` widget

ndelanou opened the pull request:
### Description
The default flinging behaviour implemented in the `Dismissible` widget often results in undesired triggering of the dismiss action when used in a scrollable content (e.g. dismissible tiles in a `ListView`)
This PR adds an optional callback parameter `shouldTriggerDismiss` to make it possible to override the default behavior.

The type of the `shouldTriggerDismiss` callback is: `bool? Function(TriggerDismissDetails details)`
When returning `null`, the default behavior is preserved and the widget will check the drag threshold and the flinging gesture.
Returning `true` will trigger the dismiss animation of the widget, even if the threshold is not reached.

The object `details` can be accessed to check for `velocity`, `progress` (relative distance dragged), `dragExtent` (distance dragged in logical pixels) or `reached` (bool value indicating whether the threshold has been reached)

### Linked issue
Closes https://github.com/flutter/flutter/issues/151223

ndelanou committed:
add param

ndelanou committed:
add test for `allowFlinging`

ndelanou committed:
fix doc

nate-thegrate commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -176,6 +177,14 @@ class Dismissible extends StatefulWidget {
+  /// flinging can dismiss an item past any threshold less than 1.0.
+  ///
+  /// This defaults to `true`
+  final bool allowFlinging;
```
The main thing here that concerns me is a potential "technical debt". Several people would benefit from a boolean flag that determines whether flinging is allowed, but there's a lot of other scenarios that people might have requests for in the future.
Maybe someone wants a fling to be successful only when it reaches a certain velocity, or maybe based on whether it reaches a certain drag extent (either a number of pixels or a percentage of the width), or maybe based on a combination of the two.

The [gestures/](https://github.com/flutter/flutter/tree/b9b0784ea4089ee394c0ba92658bd4d25e466b9d/packages/flutter/lib/src/gestures) folder has a variety of classes that store details about various gestures (a few of them are used in this widget!)

Perhaps instead of an `allowFlinging` flag, we could have something like

```dart
final bool Function(DismissGestureDetails details) shouldDismiss;
```

And then developers could choose to either provide the `dismissThresholds` map or the `shouldDismiss` callback.

ndelanou replied:
Thank you for the review. This is indeed a very good point.
I have some concerns about the potential for confusion between the new `shouldDismiss` callback and the existing `confirmDismiss` 🤔 

Also, by implementing a simple/naive check of the percentage of the width dragged in the `shouldDismiss` callback to decide if we should start a dismiss action, we will implicitly break the current behaviour of having the flinging gesture enabled by default.

nate-thegrate replied:
Thanks for bringing up the potential confusion with the `confirmDismiss` name.

Fortunately, I don't think it would need to break current behavior. Maybe something like:

```dart
class _DismissibleState extends State<Dismissible> {
  bool shouldDismiss(DismissGestureDetails details) {
    late final _FlingGestureKind fling = _describeFlingGesture(details.velocity);

    return switch (widget.shouldDismiss) {
      final callback? => callback(details),
      null => _dragExtent >= _dismissThreshold || fling == _FlingGestureKind.forward,
    };
  }

  // ...
}
```

nate-thegrate commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -176,6 +177,14 @@ class Dismissible extends StatefulWidget {
+  /// Flinging is treated as being equivalent to dragging almost to 1.0, so
+  /// flinging can dismiss an item past any threshold less than 1.0.
+  ///
+  /// This defaults to `true`
```
```suggestion
  /// Defaults to `true`.
```

nate-thegrate commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -504,8 +513,10 @@ class _DismissibleState extends State<Dismissible> with TickerProviderStateMixin
+
     final double flingVelocity = _directionIsXAxis ? details.velocity.pixelsPerSecond.dx : details.velocity.pixelsPerSecond.dy;
-    switch (_describeFlingGesture(details.velocity)) {
+    final _FlingGestureKind flingGesture = widget.allowFlinging ? _describeFlingGesture(details.velocity) :_FlingGestureKind.none;
```
```suggestion
    final _FlingGestureKind flingGesture = widget.allowFlinging ? _describeFlingGesture(details.velocity) : _FlingGestureKind.none;
```

nate-thegrate reviewed (commented):
As far as the feature implementation, this looks fantastic. Thanks for identifying a problem, implementing a solution, and updating the test framework accordingly.

I'll hold off on hitting "approve" until we hear some thoughts from another contributor. (Or if you're down to go ahead and expand on the `allowFlinging` flag, I'd be super stoked 😄)

ndelanou committed:
feedback: update default value comment

Co-authored-by: Nate Wilson <nathan.wilson1232@gmail.com>

ndelanou committed:
feedback: fix formatting

Co-authored-by: Nate Wilson <nathan.wilson1232@gmail.com>

Piinks reviewed (commented):
Hey @ndelanou! Welcome, thank you for contributing.
I agree with Nate here. The callback approach will be a lot more future proof for folks that want to customize the behavior of the Dismissible widget. Adding some sample code that demonstrates you specific use case would also be really beneficial here.

nate-thegrate commented:
As far as naming goes, here are a few AI suggestions…

![should start dismiss](https://github.com/user-attachments/assets/c9348e30-8c85-4a85-9e43-4c69f3fe2fdb)

ndelanou committed:
use `shouldDismiss` callback

ndelanou committed:
fix check order

ndelanou committed:
fix test title

ndelanou committed:
fix doc

ndelanou committed:
fix doc

ndelanou committed:
fix doc

ndelanou committed:
clean changes

nate-thegrate commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -256,6 +279,35 @@ class DismissUpdateDetails {
+    this.progress = 0.0,
+  });
+
+  /// The direction that the dismissible is being dragged.
+  final DismissDirection direction;
+
+  /// Whether the dismiss threshold is currently reached.
+  final bool reached;
+
+  /// The offset ratio of the dismissible in its parent container.
+  ///
+  /// A value of 0.0 represents the normal position and 1.0 means the child is
+  /// completely outside its parent.
+  ///
+  /// This can be used to synchronize other elements to what the dismissible is doing on screen,
+  /// e.g. using this value to set the opacity thereby fading dismissible as it's dragged offscreen.
+  final double progress;
```
I think it'd be nice to have a `velocity` member here too, and perhaps some others:

```dart
/// (passed directly from the [DragEndDetails])
final Velocity flingVelocity;

/// The component of the velocity parallel to the [DismissDirection].
/// A fling away from the center will have a positive value.
double get velocity {
  // copy some logic over from _describeFlingGesture()
}

/// The angle between the direction of the fling and the [DismissDirection].
double get angle {
  // ...
}

/// Similar to [progress], but the value is presented in logical pixels
/// rather than as a percentage of the parent's width.
/// (We're already computing this value, so we might as well use it here!)
final double dragExtent;
```

ndelanou replied:
Didn't managed to make it work for `angle`
The values returned from `DragEndDetails` are not suffisant to compute a drag angle.
The Velocity only have a value for the axis used by DismissDirection

nate-thegrate replied:
That's unfortunate to hear, but thanks for letting me know.

With this in mind, I think it would be ideal to remove the `Velocity` object, since that information is already accessible via the `double` velocity and `DismissDirection`.

nate-thegrate commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -500,6 +552,31 @@ class _DismissibleState extends State<Dismissible> with TickerProviderStateMixin
     }
     _dragUnderway = false;
+
+    // Use value returned by `widget.shouldDismiss` if a callback is provided
+    // If the callback returns null, use the default behavior
+    if (widget.shouldDismiss case final AcceptDismissCallback shouldDismissCallback) {
+      final AcceptDismissDetails details = AcceptDismissDetails(
+        direction: _dismissDirection,
+        reached: _moveController.value > _dismissThreshold,
+        progress: _moveController.value,
+      );
+      final bool? shouldDismiss = shouldDismissCallback(details);
```
Fantastic!

And the `if-case` statements are an excellent touch 😌👌

nate-thegrate commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -30,6 +30,12 @@ typedef DismissDirectionCallback = void Function(DismissDirection direction);
 /// Used by [Dismissible.confirmDismiss].
 typedef ConfirmDismissCallback = Future<bool?> Function(DismissDirection direction);
 
+/// Signature used by [Dismissible] to give the application an opportunity to
+/// accept or deny a dismiss gesture.
+///
+/// Used by [Dismissible.shouldDismiss].
+typedef AcceptDismissCallback = bool? Function(AcceptDismissDetails details);
```
I like the choice to make this nullable—a fallback to default behavior could make the implementation much more simple :)

nate-thegrate commented on packages/flutter/test/widgets/dismissible_test.dart:
```diff
@@ -867,6 +869,48 @@ void main() {
+
+    await dismissItem(tester, 1, gestureDirection: AxisDirection.right);
+    expect(find.text('1'), findsNothing);
+  });
```
Thanks for expanding on the tests here!

nate-thegrate reviewed (commented):
It's super great to see the progress here; this is going to be a fantastic addition to the framework 👍

I left some more feedback—take a look when you have a chance!

ndelanou committed:
add doc link to example

goderbauer commented:
(triage) @ndelanou Do you still have plans to come back to this PR and address the feedback given above?

ndelanou commented:
@goderbauer I just came back from vacation.
I'll get back to it soon. Thanks for asking 👍

nate-thegrate commented:
Could you try rebasing 1 more time to get [**ci.yaml validation**](https://github.com/flutter/flutter/pull/151224/checks?check_run_id=29009373312) to pass?
![Update with rebase](https://github.com/user-attachments/assets/8bacd87f-410a-4a9c-8ad0-075dd05f3eff)

ndelanou committed:
add test for API example

ndelanou committed:
fix example error

ndelanou committed:
add `dragExtent`, `flingVelocity` and `velocity` to `AcceptDismissDetails`

ndelanou committed:
fix

ndelanou committed:
fix snippet

nate-thegrate commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -691,3 +797,10 @@ class _DismissibleState extends State<Dismissible> with TickerProviderStateMixin
   }
 }
+
+extension on DismissDirection {
+  bool get directionIsXAxis => switch (this) {
+    DismissDirection.horizontal || DismissDirection.endToStart || DismissDirection.startToEnd => true,
+    _ => false
+  };
+}
```
A couple of relevant [style guidelines](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#summary):

#### avoid `extension`

Extension methods & getters are a bit more tricky to find, both in the source code and on [api.flutter.dev](https://api.flutter.dev/).

#### avoid `_`

This applies to inline callbacks (parameter names & types should be specified even when they aren't used) and also to [enum switch expressions](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-using-if-chains-or--or--with-enum-values).

The TL;DR is: we should put every enum value in the switch expression: that way, if a new value is added in the future, we can make sure it's handled correctly everywhere the enum is used.

nate-thegrate replied:
Luckily, it's an easy fix—just add it to the enum declaration:

```dart
enum DismissDirection {
  vertical,
  horizontal,
  endToStart,
  startToEnd,
  up,
  down,
  none;

  bool get directionIsXAxis => switch (this) {
    horizontal || endToStart || startToEnd => true,
    vertical   || up || down || none       => false,
  };
}
```

nate-thegrate commented on packages/flutter/lib/src/widgets/dismissible.dart:
````diff
@@ -143,6 +157,23 @@ class Dismissible extends StatefulWidget {
+  /// )
+  /// ```
+  /// {@end-tool}
+  final AcceptDismissCallback? shouldDismiss;
````
I think the concerns you mentioned earlier about confusion between `shouldDismiss` and `confirmDismiss` are valid.

What do you think about the following name change?

```dart
// before
final AcceptDismissCallback? shouldDismiss;
class AcceptDismissDetails {}


// after
final TriggerDismissCallback? shouldTriggerDismiss;
class TriggerDismissDetails {}
```

nate-thegrate commented on examples/api/test/widgets/dismissible/dismissible.1_test.dart:
These tests look great, thank you!

nate-thegrate commented on examples/api/test/widgets/dismissible/dismissible.1_test.dart:
```diff
@@ -0,0 +1,260 @@
+    expect(find.byKey(key), expectedAboveThreshold);
+  }
+
+  group('Test Drag behaviour', () {
+
+    testWidgets(
+      '0 - Default begavior - `shouldDismiss: null`',
+      (WidgetTester tester) async {
+        await initWidget(tester);
+
+        const ValueKey<int> key = ValueKey<int>(0);
+
+        await testDragBehavior(tester, key: key);
+      },
+    );
```
Suggestion for tests in this file, to match spelling/whitespace conventions elsewhere in the framework:

```suggestion
  group('Test Drag behavior', () {
    testWidgets('0 - Default behavior - `shouldDismiss: null`', (WidgetTester tester) async {
      await initWidget(tester);

      const ValueKey<int> key = ValueKey<int>(0);
      await testDragBehavior(tester, key: key);
    });
```

nate-thegrate reviewed (commented):
I'm super impressed by the progress made recently, thanks for the several improvements made here!

Just have a few more things to go over before we land this change.

ndelanou committed:
"behavior" spelling

ndelanou committed:
better formating + whitespace

ndelanou committed:
rename `shouldDismiss` by `shouldTriggerDismiss`

ndelanou committed:
remove `flingVelocity` from `TriggerDismissDetails`

ndelanou committed:
remove extension

nate-thegrate commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -259,6 +290,46 @@ class DismissUpdateDetails {
+  ///
+  /// A value of 0.0 represents the normal position and 1.0 means the child is
+  /// completely outside its parent.
+  ///
+  /// This can be used to synchronize other elements to what the dismissible is doing on screen,
+  /// e.g. using this value to set the opacity thereby fading dismissible as it's dragged offscreen.
```
```suggestion
```

That final paragraph makes sense in the context of `DismissUpdateDetails` but isn't super applicable here.

nate-thegrate commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -259,6 +290,46 @@ class DismissUpdateDetails {
+  /// e.g. using this value to set the opacity thereby fading dismissible as it's dragged offscreen.
+  final double progress;
+
+  /// Distance from the start of the drag to the current position
+  ///
+  /// The drag distance is represented in logical pixels.
+  final double dragExtent;
+
+  /// The component of the velocity parallel to the [DismissDirection].
+  /// A fling away from the center will have a positive value.
+  final double velocity;
```
tiny whitespace/wording nit:

```suggestion
  /// Distance from the start of the drag to the current position,
  /// represented in logical pixels.
  final double dragExtent;

  /// The component of the velocity parallel to the [DismissDirection].
  ///
  /// A fling away from the center will have a positive value.
  final double velocity;
```

nate-thegrate commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -335,11 +406,10 @@ class _DismissibleState extends State<Dismissible> with TickerProviderStateMixin
-  }
+  bool get _directionIsXAxis => switch (widget.direction) {
+    DismissDirection.horizontal || DismissDirection.endToStart || DismissDirection.startToEnd => true,
+    DismissDirection.vertical || DismissDirection.up || DismissDirection.down || DismissDirection.none => false
```
Thanks for converting from a `==` chain to an exhaustive switch expression!

<br>

Might as well add a [trailing comma](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#use-a-trailing-comma-for-arguments-parameters-and-list-items-but-only-if-they-each-have-their-own-line) here:

```suggestion
    DismissDirection.vertical || DismissDirection.up || DismissDirection.down || DismissDirection.none => false,
```

nate-thegrate commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -503,11 +573,41 @@ class _DismissibleState extends State<Dismissible> with TickerProviderStateMixin
+
+    final double flingVelocity = _directionIsXAxis ? details.velocity.pixelsPerSecond.dx : details.velocity.pixelsPerSecond.dy;
+
+    // Use value returned by `shouldTriggerDismiss` if a callback is provided
+    // If the callback returns null, use the default behavior
```
```suggestion
    // Use value returned by `shouldTriggerDismiss` if a callback is provided.
    // If the callback returns null, use the default behavior.
```

One [style guideline](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#use-correct-grammar) is "end all sentences with a period."

nate-thegrate commented on packages/flutter/test/widgets/dismissible_test.dart:
```diff
@@ -867,6 +869,48 @@ void main() {
+
+    await dismissItem(tester, 1, gestureDirection: AxisDirection.right);
+    expect(find.text('1'), findsOneWidget);
+  });
+  testWidgets('Use `shouldTriggerDismiss` to disable flinging', (WidgetTester tester) async {
```
nit: add blank line

```suggestion
  });

  testWidgets('Use `shouldTriggerDismiss` to disable flinging', (WidgetTester tester) async {
```

nate-thegrate reviewed (approved):
LGTM, thanks for the high-quality contribution!

(After doing another scan through the PR, I have a bunch of super tiny suggestions, so take a look when you can 🙂)

ndelanou committed:
fix typos

ndelanou committed:
remove unrelated comment

ndelanou committed:
add missing trailing comma

ndelanou committed:
Update packages/flutter/lib/src/widgets/dismissible.dart

Co-authored-by: Nate Wilson <nathan.wilson1232@gmail.com>

ndelanou committed:
Update packages/flutter/lib/src/widgets/dismissible.dart

Co-authored-by: Nate Wilson <nathan.wilson1232@gmail.com>

ndelanou committed:
Update packages/flutter/lib/src/widgets/dismissible.dart

Co-authored-by: Nate Wilson <nathan.wilson1232@gmail.com>

ndelanou committed:
Update packages/flutter/lib/src/widgets/dismissible.dart

Co-authored-by: Nate Wilson <nathan.wilson1232@gmail.com>

ndelanou committed:
Update packages/flutter/test/widgets/dismissible_test.dart

Co-authored-by: Nate Wilson <nathan.wilson1232@gmail.com>

Piinks commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -259,6 +295,43 @@ class DismissUpdateDetails {
+/// See also:
+///
+///   * [Dismissible.shouldTriggerDismiss], which receives this information.
+class TriggerDismissDetails {
```
Do we need a new class? Could we add what we need to DismissUpdateDetails be used instead for this use case? 
If not, we should definitely document what the difference is between them. :)

ndelanou replied:
`DismissUpdateDetails` contains a `bool previousReached` that makes no sense for our `shouldTriggerDismiss` callback as it is only called once per dismiss gesture (onDragEnd). This was the main reason for creating a new class.

Also, I'm adding a new `bool isFling` parameter to `TriggerDismissDetails`, that indicates if the fling velocity is enough to be recognise as a fling gesture, the same way `reached` set to true when the threshold is passed. Using `shouldTriggerDismiss: (details) => details.reached || details.isFling` should be the same as returning null (default behaviour).

I do agree that those two classes have a lot in common tho 🤔

Piinks replied:
Perhaps internal to the class we can have an enum that differentiates which scenario the details are representing? And non-applicable properties can be null then?

nate-thegrate replied:
I think my preference would be to have 2 distinct types, rather than requiring the developer to keep track of which fields are guaranteed to be null/non-null at the respective times.

A while ago, I learned that class inheritance can [reduce compile sizes](https://github.com/dart-lang/language/issues/3780#issuecomment-2094969955). So maybe we could have something like:

```dart
sealed class _DismissGestureDetails {
  final DismissDirection direction;
  final bool reached;
  final double progress;
}

class DismissUpdateDetails extends _DismissGestureDetails {
  final bool previousReached;
}

class TriggerDismissDetails extends _DismissGestureDetails {
  final double dragExtent;
  final double velocity;
}
```

Piinks commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -105,6 +118,7 @@ class Dismissible extends StatefulWidget {
     this.confirmDismiss,
+    this.shouldTriggerDismiss,
     this.onResize,
     this.onUpdate,
```
Can onUpdate not be used for this? It is called when the dismissible is dragged. I am reading through this widget (having not worked on it a lot before) and noticing we have a lot of different callbacks. If we can repurpose onUpdate I think it could be clearer.

Piinks replied:
What if we added a method for folks to programmatically trigger the dismiss instead? Then here in onUpdate folks could trigger it if they wanted. It would be a smaller amount of public API, and maybe better handle possible conflicts with other parts of the API that already exist.

ndelanou replied:
`onUpdate` is currently called using a listener on the AnimationController, so it is called multiple times when dragging the widget. It is not part of any logic right now, it's only reacting to animation changes.
`shouldTriggerDismiss` is only called once per dismiss attempt, when the drag gesture ends (where the flinging detection logic is also implemented)

While it seems like a good idea to add a return value to `onUpdate`, I don't think we will be able to tell precisely if the drag has ended when calling the `onUpdate` callback. Also, `onUpdate` is usually called at least one more time after `shouldTriggerDismiss` is called, when the animation resets or goes out of screen (dismiss).

With that in mind, I would also keep `shouldTriggerDismiss`. I don't think we can make it work with a single callback

Piinks commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -335,11 +408,10 @@ class _DismissibleState extends State<Dismissible> with TickerProviderStateMixin
-        || widget.direction == DismissDirection.endToStart
-        || widget.direction == DismissDirection.startToEnd;
-  }
+  bool get _directionIsXAxis => switch (widget.direction) {
```
The `=>` should be removed here in favor of a code block since it is multiline.

nate-thegrate replied:
[un-resolving this convo, so that hopefully the getter is turned into a block body.]

Piinks commented on packages/flutter/lib/src/widgets/dismissible.dart:
```diff
@@ -441,7 +513,7 @@ class _DismissibleState extends State<Dismissible> with TickerProviderStateMixin
 
   void _handleDismissUpdateValueChanged() {
-    if (widget.onUpdate != null) {
+    if (widget.onUpdate case final DismissUpdateCallback onUpdate) {
```
Is this change necessary?

ndelanou replied:
Switching back to a simple null check.
I'm also some kind of an "if-case" evangelist myself 😌

ndelanou replied:
I did this to reflect what I did with `shouldTriggerDismiss`

@Piinks @nate-thegrate should I also change this part to match with the current preferences on pattern matching usage within the framework ?
https://github.com/ndelanou/flutter/blob/005d193c8d79e1f069fc2056eee8a33daac40e81/packages/flutter/lib/src/widgets/dismissible.dart#L583

nate-thegrate replied:
Yeah, should probably change that one too.

Piinks reviewed (commented):
Thanks for your patience and hard work on this @ndelanou! Reading it more thoroughly, I have some questions. :)

How does this change affect dismissThresholds? If I specify a threshold there, but then the shouldTriggerDismiss has a different opinion, which should win?

nate-thegrate commented:
> How does this change affect dismissThresholds? If I specify a threshold there, but then the shouldTriggerDismiss has a different opinion, which should win?

As it is currently: if `shouldTriggerDismiss` is null, or if its return value is null, the widget defers to default behavior (based on whether the dismiss threshold was reached and/or whether a "fling" was detected).

When creating the `shouldTriggerDismiss` callback, the developer can check whether the threshold has been reached using `details.reached`.

ndelanou committed:
feedback: revert change on onUpdate null check

ndelanou committed:
feedback: get rid of "=>" return

Piinks commented:
I think there is some really good discussion here. I want to bring in another reviewer because I think there is a high chance we could be creating some confusion in the API with this new feature and the ones that already exist.

ndelanou committed:
add `isFling` parameter to `TriggerDismissDetails`

Piinks reviewed (commented):
Chatting with Nate about this today, I feel like we have a good path forward.

This is an awesome idea, love it: https://github.com/flutter/flutter/pull/151224#discussion_r1735235536

And then, how about this, we can deprecate the dismiss threshold map, since folks could achieve the same thing in the new callback. What do we think?

ndelanou committed:
create parent class for dismiss events

ndelanou commented:
> And then, how about this, we can deprecate the dismiss threshold map, since folks could achieve the same thing in the new callback. What do we think?

Well, I just came across an edge case where the result of `shouldTriggerDismiss` does not override the default behavior when the threshold is higher than 1. I'm adding a test over this case.
Removing the threshold map would eliminate this behaviour.

On the other hand, we still need a threshold value to compute the `reached` parameter used in `onUpdate` and `shouldTriggerDismiss` details. It is nice to be able to customise this threshold.

So, I don't think we can remove the map.

ndelanou committed:
add tests

ndelanou committed:
call `_handleMoveCompleted` when animation is completed

ndelanou committed:
intruduce breaking test

`shouldTriggerDismiss` result should be used event when threshould is
higher than 1

nate-thegrate commented:
> Chatting with Nate about this today, I feel like we have a good path forward.
> 
> This is an awesome idea, love it: [#151224 (comment)](https://github.com/flutter/flutter/pull/151224#discussion_r1735235536)

Thank you very much!

<br>

> I just came across an edge case where the result of `shouldTriggerDismiss` does not override the default behavior when the threshold is higher than 1. I'm adding a test over this case.

Ah, good catch 👍

<br>

> We still need a threshold value to compute the `reached` parameter used in `onUpdate` and `shouldTriggerDismiss` details. It is nice to be able to customise this threshold.
> 
> So, I don't think we can remove the map.

I agree that it's nice to be able to customize the threshold at which the dismiss is triggered, but this can be done without needing to reference `isReached`, for example:

```dart
shouldTriggerDismiss: (details) => details.progress >= 0.4 || details.isFling;
```

We talked about this pull request in a triage meeting today, and the overall feeling is that this is a highly valuable addition to the API; at the same time, having a whole bunch of configuration options for `Dismissible` can be overwhelming & confusing.

With that in mind, I think deprecating & removing the `dismissThresholds` map is the best way forward.


Feel free to read the [Handling breaking changes](https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#handling-breaking-changes) section of the Tree Hygiene wiki page for full information.

Factoring the threshold map out of the `_DismissibleState` logic is going to be tough, but the great news is that we don't have to do it! (And fortunately, you're off the hook for [data-driven fixes](https://github.com/flutter/flutter/blob/master/docs/contributing/Data-driven-Fixes.md#data-driven-fixes) as well, since they only apply to certain types of name/parameter changes.)

<br>

#### Deprecation to-do list

- [ ] Add [`@Deprecated()` decorators](https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#deprecations) in appropriate places
  - Includes `dismissThresholds` / `reached` / `previousReached`, and possibly `onUpdate`
- [ ] Move `reached` back to `DismissUpdateDetails`
  - no need to introduce the new `TriggerDismissDetails` class with deprecated members :)
- [ ] Figure out a replacement for `previousReached`
  - I think it'd make the most sense to tweak the `DismissUpdateCallback` signature.
- [ ] Add a migration guide
  1. Fork the [website repo](https://github.com/flutter/website)
  2. Make a copy of the [template file](https://github.com/flutter/website/blob/main/src/content/release/breaking-changes/template.md) and add the appropriate info, then make a PR for it

This could be a good opportunity to give `onUpdate` a more intuitive name. Possible solution:

```dart
@Deprecated(...)
typedef DismissUpdateCallback = void Function(DismissUpdateDetails details);

typedef DismissPointerUpdateCallback = void Function(DismissUpdateDetails current, DismissUpdateDetails previous);

class Dismissible {
  // ...

  @Deprecated(...)
  final DismissUpdateCallback? onUpdate;

  final DismissPointerUpdateCallback? onPointerMoved;
}
```

<br>

@ndelanou I know this probably wasn't what you expected from trying to add a single flag to this widget, but thanks for sticking it out with us. 😄

@Piinks, what do you think about splitting this into 2 PRs:
1. Add `shouldTriggerDismiss` callback, with examples & tests
2. Deprecate `dismissThresholds`, and update related APIs

chunhtai commented:
(triage) Hi @ndelanou  may I know what is the state of this pr?

victorsanni commented:
Hi @ndelanou, can you fix the failing tests?

chunhtai commented:
Closing this pr for now, feel free to reopen once you have time to get back to this pr.
